src/helper_functions.rs

```diff
@@ -0,0 +1,200 @@
+                msg::send(winner, "", exec::value_available()).expect("Can't send reward");
+                self.start_new_game();
+            }
+            _ => self.transit_to_reveal_stage(progress_description),
+        }
+    }
+
+    pub(crate) fn handle_reveal_timout(&mut self, progress_description: &StageDescription) {
+        match progress_description.finished_players.len() {
+            0 => self.update_timestamp(),
+            _ => {
+                self.end_round();
+            }
+        }
+    }
+
+    pub(crate) fn transit_to_reveal_stage(&mut self, progress_description: &StageDescription) {
+        self.stage = GameStage::Reveal(StageDescription {
+            anticipated_players: progress_description.finished_players.clone(),
+            finished_players: Default::default(),
+        });
+        self.update_timestamp();
+    }
+
+    pub(crate) fn end_round_if_needed(&mut self) -> RevealResult {
+        if let GameStage::Reveal(reveal_description) = &self.stage {
+            if reveal_description.anticipated_players.is_empty() {
+                self.end_round()
+            } else {
+                RevealResult::Continue
+            }
+        } else {
+            panic!("It's not reveal stage")
+        }
+    }
+
+    pub(crate) fn end_round(&mut self) -> RevealResult {
+        let set_of_moves = BTreeSet::from_iter(self.player_moves.values().cloned());
+        let next_round_players: BTreeSet<ActorId> = match set_of_moves.len() {
+            1 | 4 | 5 => self.player_moves.keys().cloned().collect(),
+            2 | 3 => {
+                let winners = self.next_round_moves_set(set_of_moves);
+                self.player_moves
+                    .iter()
+                    .filter(|(_, users_move)| winners.contains(users_move))
+                    .map(|(player, _)| player)
+                    .copied()
+                    .collect()
+            }
+            _ => panic!("Unknown result"),
+        };
+
+        if next_round_players.len() > 1 {
+            self.stage = GameStage::InProgress(StageDescription {
+                anticipated_players: next_round_players.clone(),
+                finished_players: BTreeSet::new(),
+            });
+            self.update_timestamp();
+            self.clear_moves();
+
+            RevealResult::NextRoundStarted {
+                players: next_round_players,
+            }
+        } else {
+            let winner = next_round_players
+                .into_iter()
+                .last()
+                .expect("Unknown winner");
+            msg::send(winner, "WIN", exec::value_available()).expect("Can't send reward");
+            self.start_new_game();
+
+            RevealResult::GameOver { winner }
+        }
+    }
+
+    pub(crate) fn next_round_moves_set(&self, set_of_moves: BTreeSet<Move>) -> BTreeSet<Move> {
+        'outer: for a_move in set_of_moves.iter() {
+            for b_move in set_of_moves.iter() {
```

Comment:
No need to call `iter` explicitly.
```suggestion
        'outer: for a_move in &set_of_moves {
            for b_move in &set_of_moves {
```